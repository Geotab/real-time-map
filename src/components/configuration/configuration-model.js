export function selectTab(tab) {
	const selectedTab = document.getElementById(tab);

	if (selectedTab.classList.contains("RTM-TabSelected")) {
		return false;
	}
	else {

		const tabs = document.getElementsByClassName("RTM-Tab");
		for (let i = 0; i < tabs.length; i++) {
			tabs[i].classList.remove("RTM-TabSelected");
			tabs[i].classList.add("RTM-TabNotSelected");
		};

		selectedTab.classList.remove("RTM-TabNotSelected");
		selectedTab.classList.add("RTM-TabSelected");

		switch (selectedTab.id) {
			case "RTM-ExceptionTitle":
				document.getElementById("exception-tab").style.display = "block";
				document.getElementById("status-tab").style.display = "none";
				document.getElementById("vehicle-tab").style.display = "none";
				break;
			case "RTM-StatusTitle":
				document.getElementById("status-tab").style.display = "block";
				document.getElementById("exception-tab").style.display = "none";
				document.getElementById("vehicle-tab").style.display = "none";
				break;
			case "RTM-VehicleTitle":
				document.getElementById("vehicle-tab").style.display = "block";
				document.getElementById("exception-tab").style.display = "none";
				document.getElementById("status-tab").style.display = "none";
				break;
		}
	}
}
