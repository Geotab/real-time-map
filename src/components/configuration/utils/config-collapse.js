import storage from "../../../dataStore";

export const initCollapse = () => {
	const coll = document.getElementsByClassName("collapsible");

	coll[0].addEventListener("click", function() {

		this.classList.toggle("active");
		const content = document.getElementById("RTM-config-container");
		const map = document.getElementById("RTM-Map");
		const button = document.getElementById("collapse-button");

		if (content.style.maxWidth === "100%") {
			button.title = "Open Configuration Panel";
			button.classList.remove("closeConfigPanel");
			button.classList.add("openConfigPanel");
			content.style.maxWidth = "0%";
			map.style.maxWidth = "100%";
		}
		else {
			button.title = "Collapse Configuration Panel";
			button.classList.remove("openConfigPanel");
			button.classList.add("closeConfigPanel");
			content.style.maxWidth = "100%";
			map.style.maxWidth = "75%";
		}

		storage.map.invalidateSize(true);
	});
};
