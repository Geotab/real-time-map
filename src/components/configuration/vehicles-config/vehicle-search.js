import { fromEvent } from "rxjs";
import { map, debounceTime } from "rxjs/operators";
import { markerList } from "../../../dataStore/map-data";
import { getDevicesInGroups, createGroupsByNameCall, createDeviceByNameCall, makeAPIMultiCall, getBlobStorage, saveBlobStorage, setBlobStorage } from "../../../services/api/helpers";
import { showSnackBar } from "../../../components/snackbar/snackbar";
import storage from "../../../dataStore";
import layerModel from "../../map/layers";

export const deviceSearch = {
   shown: true,
   deviceResultsCache: {},
   selectedIDS: {},

   get searchInput() {
      return document.getElementById("RTM-vehicle-search-bar");
   },

   init(mapPropsToComponent) {
      // Init rxjs debounce search.
      const searchInputObservable = fromEvent(deviceSearch.searchInput, "input").pipe(map(i => i.currentTarget.value));
      const debouncedInput = searchInputObservable.pipe(debounceTime(250));
      debouncedInput.subscribe((searchInput) => deviceSearch.buildSearchList(searchInput, mapPropsToComponent));
   },

   loadSavedDeviceConfig(mapPropsToComponent) {

      return getBlobStorage().then(val => {
         if (val.length === 0) { return; }
         const cachedDevices = JSON.parse(val[0].data);

         if (cachedDevices.configData.Vehicle) {
            deviceSearch.selectedIDS = cachedDevices.configData.Vehicle;
            deviceSearch.applyFilter();
            deviceSearch.buildDeviceDisplayList(mapPropsToComponent);
         }
      });
   },

   buildSearchList(searchInput, mapPropsToComponent) {

      const nameSearchMultiCall = [
         createDeviceByNameCall(searchInput),
         createGroupsByNameCall(searchInput)
      ];

      return makeAPIMultiCall(nameSearchMultiCall).then(results => {
         // deviceList = results[0];
         results[0]
            .map(deviceSearch.saveDeviceDataToCache);
         // groupList = results[1];
         results[1]
            .map(deviceSearch.saveGroupDataToCache);
         mapPropsToComponent(Object.values(deviceSearch.deviceResultsCache));
      });

   },

   saveDeviceDataToCache(deviceData) {
      const data = {};
      ["id", "name", "groups"].forEach(prop => data[prop] = deviceData[prop]);
      data.visible = true;
      deviceSearch.deviceResultsCache[data.name] = data;
      return data;
   },

   saveGroupDataToCache(deviceData) {
      const data = {};
      ["id", "name", "color"].forEach(prop => data[prop] = deviceData[prop]);
      data.visible = true;
      data.name += " (Group)";

      deviceSearch.deviceResultsCache[data.name] = data;
      return data;
   },

   buildDeviceDisplayList(mapPropsToComponent) {
      deviceSearch.mapPropsToComponent = mapPropsToComponent;
      mapPropsToComponent(Object.values(deviceSearch.selectedIDS)
         .filter(device => device.id && device.name));
   },

   handleItemSelected(event, mapPropsToComponent) {
      event.preventDefault();
      deviceSearch.saveSelectedValue(mapPropsToComponent);
   },

   saveSelectedValue(mapPropsToComponent) {

      const { value } = deviceSearch.searchInput;
      if (!deviceSearch.deviceResultsCache.hasOwnProperty(value)) {
         return;
      }

      const deviceData = deviceSearch.deviceResultsCache[value];
      deviceData.visible = true;
      const {
         id,
         name,
         groups,
         color,
         visible
      } = deviceData;

      if (color) { //It's a group
         deviceSearch.selectedIDS[id] = { id, name, color, visible };
      }
      else {
         deviceSearch.selectedIDS[id] = { id, name, groups, visible };
      }

      deviceSearch.searchInput.value = "";
      deviceSearch.saveConfig(mapPropsToComponent);
      return deviceData;
   },

   deleteItemFromdeviceList(id, mapPropsToComponent) {
      delete deviceSearch.selectedIDS[id];
      deviceSearch.saveConfig(mapPropsToComponent);
   },

   deleteAllItems(mapPropsToComponent) {
      deviceSearch.selectedIDS = {};
      deviceSearch.saveConfig(mapPropsToComponent);
   },

   saveConfig(mapPropsToComponent) {
      storage.setBlobStorageObj ? setBlobStorage("Vehicle", deviceSearch.selectedIDS) : saveBlobStorage("Vehicle", deviceSearch.selectedIDS);

      deviceSearch.applyFilter();
      deviceSearch.buildDeviceDisplayList(mapPropsToComponent);
   },

   applyFilter() {
      _getDeviceList(deviceSearch.selectedIDS).then(() => {
         _applyDeviceFilter(Object.keys(storage.selectedDevices));
      });
   },
   showAll(mapPropsToComponent) {
      Object.values(deviceSearch.selectedIDS)
         .forEach(selectedDevice =>
            selectedDevice.visible = true
         );
      deviceSearch.saveConfig(mapPropsToComponent);
   },
   hideAll(mapPropsToComponent) {
      Object.values(deviceSearch.selectedIDS)
         .forEach(selectedDevice =>
            selectedDevice.visible = false
         );
      deviceSearch.saveConfig(mapPropsToComponent);
   },
   toggleDeviceVisibility(id, mapPropsToComponent) {
      const selectedDevice = deviceSearch.selectedIDS[id];
      selectedDevice.visible = !selectedDevice.visible;
      deviceSearch.saveConfig(mapPropsToComponent);
   },
   zoomIntoDevice(id) {
      const deviceMarker = markerList[id];
      if (deviceMarker) {
         const newZoomLevel = Math.max(Math.min(storage.map.getZoom() + 1, 18), 15);
         storage.map.flyTo(deviceMarker.currentlatLng, newZoomLevel);
      } else {
         showSnackBar("Sorry, no current day data for selected vehicle.");
      }
   }
};

export function _getDeviceList(selectedIDs) {
   storage.selectedDevices = {};
   return Promise.all(Object.values(selectedIDs).map(data => {
      const {
         id,
         color,
         visible
      } = data;
      if (!visible) {
         return Promise.resolve("done");
      }

      if (color) { //It's a group
         return getDevicesInGroups([{ id }]).then(devices =>
            devices.forEach(device => {
               const {
                  id,
                  name,
                  groups
               } = device;
               storage.selectedDevices[id] = { id, name, groups, visible: true };
            })
         );

      }
      else {
         storage.selectedDevices[id] = data;
         return Promise.resolve("done");
      }
   }));
}

window.addDeviceToFilter = (id, name = "Go Device") => {
   deviceSearch.selectedIDS[id] = { id, name, visible: true };
   deviceSearch.saveConfig(deviceSearch.mapPropsToComponent);
   openPopup();
};

export function _applyDeviceFilter(deviceIDS) {
   // console.log(deviceIDS);
   layerModel.hideAllLayers();

   if (deviceIDS.length === 0) {
      layerModel.showLayer("movingLayer");
      return;
   }

   const layerName = "Filter";

   if (!layerModel.layerList.hasOwnProperty(layerName)) {
      layerModel.createNewLayer(layerName);
   }

   layerModel.layerList[layerName].clearLayers();
   layerModel.showLayer(layerName);

   deviceIDS.forEach(deviceID => {
      const deviceMarker = markerList[deviceID];
      if (deviceMarker) {
         deviceMarker.setLayer(layerName);
      }
   });
}

function openPopup() {
   const collapseButton = document.getElementsByClassName("collapsible");
   const vehicleTab = document.getElementById("RTM-VehicleTitle");
   if (!collapseButton[0].classList.contains("active")) {
      collapseButton[0].click();
      vehicleTab.click();
   }
}
