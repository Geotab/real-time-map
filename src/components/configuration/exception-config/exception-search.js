import { fromEvent } from "rxjs";
import { map, debounceTime } from "rxjs/operators";
import { getRulesByName, getBlobStorage, saveBlobStorage, setBlobStorage } from "../../../services/api/helpers";
import storage from "../../../dataStore";
import { filterByVisibility } from "../utils/config-helpers";

export const exceptionSearch = {
	exceptionResultsCache: {},
	displayList: {},

	get searchInput() {
		return document.getElementById("RTM-exception-search-bar");
	},

	init(mapPropsToComponent) {
		// Init rxjs debounce search.
		const searchInputObservable$ = fromEvent(exceptionSearch.searchInput, "input").pipe(map(i => i.currentTarget.value));
		const debouncedInput = searchInputObservable$.pipe(debounceTime(250));
		debouncedInput.subscribe((searchInput) => exceptionSearch.buildSearchList(searchInput, mapPropsToComponent));
	},

	loadSavedExceptionConfig(mapPropsToComponent) {

		return getBlobStorage().then(val => {

			if (val.length > 0 && !storage.setBlobStorageObj) { storage.setBlobStorageObj = val[0]; }
			else {
				return saveBlobStorage("Exception", {}).then(() => {
					return getBlobStorage().then(val => {
						storage.setBlobStorageObj = val[0];
					});
				});
			};

			exceptionSearch.searchInput.value = "";
			const cachedExceptions = JSON.parse(val[0].data);

			if (cachedExceptions.configData.Exception) {
				exceptionSearch.displayList = cachedExceptions.configData.Exception;
				storage.selectedExceptions = filterByVisibility(exceptionSearch.displayList);
				exceptionSearch.buildExceptionDisplayList(mapPropsToComponent);
			}
		}).catch(error =>
			console.warn("Server is unavailable, please try again later: ", error)
		);
	},

	buildSearchList(searchInput, mapPropsToComponent) {
		return getRulesByName(searchInput).then(ruleList => {

			ruleList.forEach(rule => {
				let {
					color,
					name,
					id,
					baseType
				} = rule;

				if (baseType === "Stock") { name += " (Default)"; };
				const visible = true;

				exceptionSearch.exceptionResultsCache[name] = { color, name, id, visible, baseType };
			});

			mapPropsToComponent(Object.values(exceptionSearch.exceptionResultsCache));
		});
	},

	handleItemSelected(event, mapPropsToComponent) {
		event.preventDefault();
		exceptionSearch.saveSelectedValue(mapPropsToComponent);
	},

	saveSelectedValue(mapPropsToComponent) {
		const { value } = exceptionSearch.searchInput;
		const exceptionData = exceptionSearch.exceptionResultsCache[value];

		if (exceptionData) {
			exceptionSearch.displayList[exceptionData.id] = exceptionData;
			exceptionSearch.saveConfig(mapPropsToComponent);
			exceptionSearch.searchInput.value = "";
		}
		return exceptionData;
	},

	buildExceptionDisplayList(mapPropsToComponent) {
		mapPropsToComponent(Object.values(exceptionSearch.displayList)
			.filter(ruleData => ruleData.id && ruleData.name));
	},

	deleteItemFromExceptionList(id, mapPropsToComponent) {
		delete exceptionSearch.displayList[id];
		exceptionSearch.saveConfig(mapPropsToComponent);
	},

	saveConfig(mapPropsToComponent) {
		storage.selectedExceptions = filterByVisibility(exceptionSearch.displayList);
		storage.setBlobStorageObj ? setBlobStorage("Exception", exceptionSearch.displayList) : saveBlobStorage("Exception", exceptionSearch.displayList);
		exceptionSearch.buildExceptionDisplayList(mapPropsToComponent);
		storage.dateKeeper$.update();
	},

	toggleExceptionVisibility(ruleID, mapPropsToComponent) {
		const selectedException = exceptionSearch.displayList[ruleID];
		selectedException.visible = !selectedException.visible;
		exceptionSearch.saveConfig(mapPropsToComponent);
	},
	deleteAllItems(mapPropsToComponent) {
		exceptionSearch.displayList = {};
		exceptionSearch.saveConfig(mapPropsToComponent);
	},
	setVisibilityForAllItems(visibility, mapPropsToComponent) {
		Object.values(exceptionSearch.displayList)
			.forEach(selectedItem =>
				selectedItem.visible = visibility
			);
		exceptionSearch.saveConfig(mapPropsToComponent);
	}
};
