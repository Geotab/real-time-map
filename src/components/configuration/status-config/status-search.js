import { fromEvent } from "rxjs";
import { map, debounceTime } from "rxjs/operators";
import { getDiagnosticByName, getBlobStorage, saveBlobStorage } from "../../../services/api/helpers";
import storage from "../../../dataStore";
import { filterByVisibility } from "../utils/config-helpers";
import { STATUS } from "../../../constants/status-search";

export const diagnosticSearch = {
	resultsCache: {},
	displayList: {},

	get searchInput() {
		return document.getElementById("RTM-status-search-bar");
	},

	init(mapPropsToComponent) {
		// Init rxjs debounce search.
		const searchInputObservable$ = fromEvent(diagnosticSearch.searchInput, "input").pipe(map(i => i.currentTarget.value));
		const debouncedInput = searchInputObservable$.pipe(debounceTime(250));
		debouncedInput.subscribe((searchInput) => diagnosticSearch.buildSearchList(searchInput, mapPropsToComponent));
	},

	loadSavedStatusConfig(mapPropsToComponent) {

		return getBlobStorage().then(val => {

			const cachedDiagnostics = val
				.map(e => JSON.parse(e.data))
				.filter(e => e.configData.type === STATUS)
				.sort((a, b) => b.date - a.date);

			if (cachedDiagnostics.length > 0) {

				// Apply Status Data retrieved from Blob Storage
				const cachedDiagnostic = cachedDiagnostics[0].configData;
				diagnosticSearch.displayList = cachedDiagnostic.typeData;
				storage.selectedStatuses = filterByVisibility(diagnosticSearch.displayList);
				diagnosticSearch.buildStatusDisplayList(mapPropsToComponent);
			}
		});
	},

	buildSearchList(searchInput, mapPropstoComponent) {

		return getDiagnosticByName(searchInput).then(diagnosticResults => {

			diagnosticResults.forEach(diagnostic => {
				const {
					id,
					name,
					unitOfMeasure
				} = diagnostic;

				const visible = true;

				diagnosticSearch.resultsCache[name] = { name, id, visible, unitOfMeasure };
			});
			mapPropstoComponent(Object.values(diagnosticSearch.resultsCache));
		});
	},

	handleItemSelected(event, mapPropsToComponent) {
		event.preventDefault();
		diagnosticSearch.saveSelectedValue(mapPropsToComponent);
	},

	saveSelectedValue(mapPropsToComponent) {

		const { value } = diagnosticSearch.searchInput;
		const diagnosticData = diagnosticSearch.resultsCache[value];

		if (diagnosticData) {
			diagnosticSearch.displayList[diagnosticData.id] = diagnosticData;
			diagnosticSearch.searchInput.value = "";
			diagnosticSearch.saveConfig(mapPropsToComponent);
		}

		return diagnosticData;
	},

	buildStatusDisplayList(mapPropsToComponent) {
		mapPropsToComponent(Object.values(diagnosticSearch.displayList)
			.filter(diagnostic => diagnostic.id && diagnostic.name));
	},

	deleteItemFromStatusList(id, mapPropsToComponent) {
		delete diagnosticSearch.displayList[id];
		diagnosticSearch.saveConfig(mapPropsToComponent);
	},

	saveConfig(mapPropsToComponent) {
		storage.selectedStatuses = filterByVisibility(diagnosticSearch.displayList);
		saveBlobStorage("Status", diagnosticSearch.displayList);
		storage.dateKeeper$.update();
		diagnosticSearch.buildStatusDisplayList(mapPropsToComponent);
	},

	toggleStatusVisibility(id, mapPropsToComponent) {
		const selectedDiagnostic = diagnosticSearch.displayList[id];
		selectedDiagnostic.visible = !selectedDiagnostic.visible;
		diagnosticSearch.saveConfig(mapPropsToComponent);
	},

	deleteAllItems(mapPropsToComponent) {
		diagnosticSearch.displayList = {};
		diagnosticSearch.saveConfig(mapPropsToComponent);
	},

	showAllItems(mapPropsToComponent) {
		Object.values(diagnosticSearch.displayList)
			.forEach(selectedDiagnostic =>
				selectedDiagnostic.visible = true
			);
		diagnosticSearch.saveConfig(mapPropsToComponent);
	},

	hideAllItems(mapPropsToComponent) {
		Object.values(diagnosticSearch.displayList)
			.forEach(selectedDiagnostic =>
				selectedDiagnostic.visible = false
			);
		diagnosticSearch.saveConfig(mapPropsToComponent);
	}
};
