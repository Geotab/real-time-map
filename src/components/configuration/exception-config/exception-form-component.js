import React, { Component } from "react";
import { exceptionSearch } from "./exception-search";

export class ExceptionFormComponent extends React.Component {
  constructor(props) {
    super(props);
    this.changeHandler = this.changeHandler.bind(this);
  }

  changeHandler(e) {
    exceptionSearch.handleItemSelected(e, this.props.setExceptionsDisplay);
  }
  render() {
    return (
      <>
        <input
          id="RTM-exception-search-bar"
          type="search"
          placeholder="Search..."
          list="RTM-exceptions-list"
          className="RTM-SearchBar Exceptions"
          onChange={this.changeHandler}
          autoComplete="off"
        ></input>
        <datalist id="RTM-exceptions-list">
          <Dropdown options={this.props.exceptions} />
        </datalist>
      </>
    );
  }
}

export const Dropdown = props => {
  const isResultEmpty = props.options.length <= 0;
  const builtDropdownList = props.options.map(prop => (
    <option key={prop.id} id={prop.id}>
      {prop.name}
    </option>
  ));
  const optionList = isResultEmpty ? [] : builtDropdownList;
  return optionList;
};
