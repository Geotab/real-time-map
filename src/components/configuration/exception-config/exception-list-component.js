import React, { Component } from "react";
import { exceptionSearch } from "./exception-search";

export const ExceptionListComponent = props => {
  const mapPropsToComponent = props.setExceptionsDisplay;
  const exceptionList =
    props.exceptionDisplayList.length > 0
      ? props.exceptionDisplayList.map(prop => (
          <li key={prop.id} className="mdc-list-item">
            <span
              className={`RTM-iconSquare mdc-list-item__graphic material-icons filterIcon ${
                prop.visible ? "showConfig" : "hideConfig"
              } `}
              title="Hide/Show"
              onClick={() =>
                exceptionSearch.toggleExceptionVisibility(
                  prop.id,
                  mapPropsToComponent
                )
              }
            ></span>
            <span
              className="RTM-iconSquare mdc-list-item__graphic material-icons exception"
              style={{
                backgroundColor: `rgba(${prop.color["r"]},${prop.color["g"]},${prop.color["b"]})`
              }}
            ></span>
            <span
              id={"RTMnode-" + prop.id}
              className="RTM-ConfigListItem mdc-list-item__text"
            >
              {prop.name}
            </span>
            <span
              className="mdc-list-item__meta material-icons"
              onClick={() =>
                exceptionSearch.deleteItemFromExceptionList(
                  prop.id,
                  mapPropsToComponent
                )
              }
            ></span>
          </li>
        ))
      : [];
  return exceptionList;
};
