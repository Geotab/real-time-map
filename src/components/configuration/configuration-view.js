import React, { Component } from "react";
import { ExceptionsTab } from "./exception-config/exceptions-tab";
import { StatusTab } from "./status-config/status-tab";
import { VehiclesTab } from "./vehicles-config/vehicles-tab";
import { initCollapse } from "./utils/config-collapse";
import { diagnosticSearch } from "./status-config/status-search";
import { exceptionSearch } from "./exception-config/exception-search";
import { deviceSearch } from "./vehicles-config/vehicle-search";

export class ConfigView extends React.Component {
   constructor(props) {
      super(props);
      this.state = {
         exceptionsSearchList: [],
         statuses: [],
         devices: [],
         exceptionDisplayList: [],
         statusDisplayList: [],
         vehicleDisplayList: []
      };
      this.setExceptions = this.setExceptions.bind(this);
      this.setExceptionsList = this.setExceptionsList.bind(this);
      this.setStatusList = this.setStatusList.bind(this);
      this.setVehicleList = this.setVehicleList.bind(this);
      this.setStatuses = this.setStatuses.bind(this);
      this.setDevices = this.setDevices.bind(this);
   }

   setExceptions(props) {
      this.setState({ exceptionsSearchList: props });
   }

   setExceptionsList(props) {
      this.setState({ exceptionDisplayList: props });
   }

   setStatusList(props) {
      this.setState({ statusDisplayList: props });
   }

   setVehicleList(props) {
      this.setState({ vehicleDisplayList: props });
   }

   setStatuses(props) {
      this.setState({ statuses: props });
   }

   setDevices(props) {
      this.setState({ devices: props });
   }

   componentDidMount() {
      exceptionSearch.init(this.setExceptions);
      diagnosticSearch.init(this.setStatuses);
      deviceSearch.init(this.setDevices);
      exceptionSearch.loadSavedExceptionConfig(this.setExceptionsList);
      diagnosticSearch.loadSavedStatusConfig(this.setStatusList);
      deviceSearch.loadSavedDeviceConfig(this.setVehicleList);
      initCollapse();
   }

   render() {
      return (
         <div id="RTM-config-view">
            <div id="RTM-config-container">
               <div id="RTM-config-header" className="">
                  <VehiclesTab
                     devices={this.state.devices}
                     vehicleDisplayList={this.state.vehicleDisplayList}
                     onClick={this.setVehicleList}
                  />
                  <StatusTab
                     statuses={this.state.statuses}
                     statusDisplayList={this.state.statusDisplayList}
                     onClick={this.setStatusList}
                  />
                  <ExceptionsTab
                     exceptions={this.state.exceptionsSearchList}
                     exceptionDisplayList={this.state.exceptionDisplayList}
                     onClick={this.setExceptionsList}
                  />
               </div>
            </div>
         </div>
      );
   }
}
