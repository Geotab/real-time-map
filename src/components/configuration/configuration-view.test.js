import React from "react";
import { ConfigView } from "./configuration-view";
import { render, shallow } from "enzyme";
import {
  spyOnFunction,
  createDivsWithID
} from "../../../tests/utils/test-helpers";

import * as collapse from "./utils/config-collapse";
import { diagnosticSearch } from "./status-config/status-search";
import { exceptionSearch } from "./exception-config/exception-search";
import { deviceSearch } from "./vehicles-config/vehicle-search";
import { selectTab } from "./configuration-model";

describe("ConfigView Tests", () => {

  afterAll(() => {
    jest.restoreAllMocks();
  });

  test("should successfully render the config component", () => {
    const component = render(<ConfigView />);
    expect(component).toMatchSnapshot();
  });

  test("should successfully call the component did mount when rendering view", () => {
    spyOnFunction(diagnosticSearch, "init", fn => fn());
    spyOnFunction(deviceSearch, "init", fn => fn());
    spyOnFunction(exceptionSearch, "init", fn => fn());
    spyOnFunction(collapse, "initCollapse");
    spyOnFunction(deviceSearch, "loadSavedDeviceConfig", fn => fn());
    spyOnFunction(exceptionSearch, "loadSavedExceptionConfig", fn => fn());
    spyOnFunction(diagnosticSearch, "loadSavedStatusConfig", fn => fn());

    shallow(<ConfigView />);

    expect(collapse.initCollapse).toHaveBeenCalled();
    expect(deviceSearch.loadSavedDeviceConfig).toHaveBeenCalled();
    expect(exceptionSearch.loadSavedExceptionConfig).toHaveBeenCalled();
    expect(diagnosticSearch.loadSavedStatusConfig).toHaveBeenCalled();
  });
});

describe("Configuration-model Tests", () => {

  const testTabID = "testTabID";

  afterAll(() => {
    jest.restoreAllMocks();
  });

  test("selectTab", () => {
    document.body.innerHTML = `<div id='${testTabID}' class="RTM-TabSelected"></div>`;
    const result = selectTab(testTabID);
    expect(result).toBe(false);
  });

  test("selectTab", () => {

    const exceptionTitleID = "RTM-ExceptionTitle";
    document.body.innerHTML = `<div id='${exceptionTitleID}' class="RTM-Tab"></div>`;
    createDivsWithID(["exception-tab", "status-tab", "vehicle-tab"]);

    selectTab(exceptionTitleID);
  });


  test("selectTab", () => {

    const statusTitleID = "RTM-StatusTitle";
    document.body.innerHTML = `<div id='${statusTitleID}' class="RTM-Tab"></div>`;
    createDivsWithID(["exception-tab", "status-tab", "vehicle-tab"]);

    selectTab(statusTitleID);
  });

  test("selectTab", () => {

    const vehicleTitleID = "RTM-VehicleTitle";
    document.body.innerHTML = `<div id='${vehicleTitleID}' class="RTM-Tab"></div>`;
    createDivsWithID(["exception-tab", "status-tab", "vehicle-tab"]);

    selectTab(vehicleTitleID);
  });
});