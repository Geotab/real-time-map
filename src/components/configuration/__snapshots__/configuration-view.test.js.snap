// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigView Tests should successfully render the config component 1`] = `
<div
  id="RTM-config-view"
>
  <div
    id="RTM-config-container"
  >
    <div
      class=""
      id="RTM-config-header"
    >
      <div>
        <button
          class="RTM-Tab RTM-TabSelected"
          id="RTM-ExceptionTitle"
        >
          Exceptions
        </button>
        <div
          class="RTM-config-info"
          id="exception-tab"
        >
          <input
            autocomplete="off"
            class="RTM-SearchBar Exceptions"
            id="RTM-exception-search-bar"
            list="RTM-exceptions-list"
            placeholder="Search..."
            type="search"
          />
          <datalist
            id="RTM-exceptions-list"
          />
          <div
            class="config-Header"
          >
            <p>
              Selected Exceptions:
            </p>
            <div
              class="toggleButtonDiv"
              id="ToggleExceptionButtons"
            >
              <div>
                <button
                  class="toggleButton shown"
                  id="clearExceptions"
                  title="Toggle All Exceptions"
                />
                <button
                  class="deleteButton"
                  id="deleteExceptions"
                  title="Delete All Exceptions"
                />
              </div>
            </div>
          </div>
          <div
            id="RTM-exceptions-view"
          >
            <div
              class="config-list"
            >
              <div>
                <ul
                  class="mdc-list"
                  id="ExceptionsList"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <button
          class="RTM-Tab RTM-TabNotSelected"
          id="RTM-StatusTitle"
        >
          Status
        </button>
        <div
          class="RTM-config-info"
          id="status-tab"
        >
          <input
            autocomplete="off"
            class="RTM-SearchBar Status"
            id="RTM-status-search-bar"
            list="RTM-status-list"
            placeholder="Search..."
            type="search"
          />
          <datalist
            id="RTM-status-list"
          />
          <div
            class="config-Header"
          >
            <p>
              Selected Status:
            </p>
            <div
              class="toggleButtonDiv"
              id="ToggleStatusButtons"
            >
              <button
                class="toggleButton shown"
                id="toggleStatus"
                title="Toggle All Statuses"
              />
              <button
                class="deleteButton"
                id="deleteStatus"
                title="Delete All Statuses"
              />
            </div>
          </div>
          <div
            id="RTM-status-view"
          >
            <div
              class="config-list"
            >
              <div>
                <ul
                  class="mdc-list"
                  id="StatusList"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <button
          class="RTM-Tab RTM-TabNotSelected"
          id="RTM-VehicleTitle"
        >
          Vehicles
        </button>
        <div
          class="RTM-config-info"
          id="vehicle-tab"
        >
          <input
            autocomplete="off"
            class="RTM-SearchBar Vehicles"
            id="RTM-vehicle-search-bar"
            list="RTM-vehicles-list"
            placeholder="Search..."
            type="search"
          />
          <datalist
            id="RTM-vehicles-list"
          />
          <div
            class="config-Header"
          >
            <p>
              Selected Vehicles:
            </p>
            <div
              class="toggleButtonDiv"
              id="ToggleDeviceButtons"
            >
              <button
                class="toggleButton shown"
                id="clearDevices"
                title="Toggle All Devices/Groups"
              />
              <button
                class="deleteButton"
                id="deleteDevices"
                title="Delete All Devices/Groups"
              />
            </div>
          </div>
          <div
            id="RTM-vehicles-view"
          >
            <div
              class="config-list"
            >
              <div>
                <ul
                  class="mdc-list"
                  id="VehicleList"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
